### PR TITLE
*: support parsing BACKUP and RESTORE statements

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Added the BACKUP, RESTORE and SHOW BACKUP/RESTORE statements to parser.

```sql
BACKUP DATABASE * TO 'local:///tmp/archive01/';
BACKUP DATABASE `a` TO 'local:///tmp/archive02/';
BACKUP TALBE `a`.`b` TO 'local:///tmp/archive03/';
RESTORE DATABASE * FROM 'local:///tmp/archive01/';
RESTORE DATABASE `a` FROM 'local:///tmp/archive02/';
RESTORE TALBE `a`.`b` FROM 'local:///tmp/archive03/';
BACKUP DATABASE * FULL TO 'local:///tmp/archive04/';
BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP '2020-02-02 14:15:16' TO 'local:///tmp/archive05/';
BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP_ORACLE 1234567890 TO 'local:///tmp/archive06/';

BACKUP DATABASE * TO 's3://my_bucket/path' 
    RATE_LIMIT = 100 MB/SECOND 
    SNAPSHOT = 2 HOUR AGO 
    CONCURRENCY = 32 
    S3_REGION = 'us-west-1' 
    S3_USE_ACCELERATE_ENDPOINT = 1;

SHOW BACKUP;
SHOW RESTORE;
SHOW BACKUP LIKE 'b01234';
SHOW BACKUP WHERE status <> 'running';
```

These are needed for integrating BR into TiDB (PR will come later).

(Note: because the new unresolved keyword `S3_USE_ACCELERATE_ENDPOINT` is longer than all existing tokens, forcing `gofmt` to replace the entire token map, I've taken advantage of this to sort the entire tokenMap)

### What is changed and how it works?

↑

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: parser.y [hunk 1]
```diff
@@ -4020,6 +4051,209 @@ ExplainFormatType:
...
+		}
+		// TODO: check overflow?
+		$$ = &ast.BRIEOption{
+			Tp:        ast.BRIEOptionBackupTimeAgo,
+			UintValue: $3.(uint64) * uint64(unit),
+		}
+	}
+|	"RATE_LIMIT" EqOpt LengthNum "MB" '/' "SECOND"
+	{
+		// TODO: check overflow?
```
Comment (tangenta) [inline parser.y]: How can it overflow :grey_question:

File: parser_test.go [hunk 2]
```diff
@@ -4969,3 +4973,50 @@ func (s *testParserSuite) TestIndexAdviseStmt(c *C) {
...
+
+		{"BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP '2020-02-02 14:14:14' TO 'noop://'", true, "BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP '2020-02-02 14:14:14' TO 'noop://'"},
+		{"BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP_ORACLE 1234567890 TO 'noop://'", true, "BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP_ORACLE 1234567890 TO 'noop://'"},
+		{"BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP_ORACLE '2020-02-02 14:14:14' TO 'noop://'", false, ""},
+		{"BACKUP DATABASE * INCREMENTAL UNTIL TIMESTAMP 1234567890 TO 'noop://'", false, ""},
+
+		{"backup database * to 'noop://' rate_limit 500 MB/second snapshot 5 minute ago", true, "BACKUP DATABASE * TO 'noop://' RATE_LIMIT = 500 MB/SECOND SNAPSHOT = 300000000 MICROSECOND AGO"},
+		{"restore table g from 'noop://' concurrency 40 checksum 0 online 1", true, "RESTORE TABLE `g` FROM 'noop://' CONCURRENCY = 40 CHECKSUM = 0 ONLINE = 1"},
+		{
+			// FIXME: should we really include the access key in the Restore() text???
```
Comment (tangenta) [inline parser_test.go]: Should we also handle `SetPwdStmt.Restore()`?

File: parser.y (hunk 1 above)
Comment (kennytm) [reply parser.y]: @tangenta `RATE_LIMIT = 9223372036854775807 MB/SECOND`... admittedly this is not a real issue 😅

File: parser_test.go (hunk 2 above)
Comment (kennytm) [reply parser_test.go]: 😱. Perhaps deal with this later (e.g. add a `WriteSecureString()` method). This needs to be considered together with how TiDB uses `Restore()`.

**Comment by tiancaiamao** [comment]
LGTM
With this feature implemented,  preparing some test cases can be much faster.

File: ast/misc.go [hunk 3]
```diff
@@ -2196,6 +2196,207 @@ func (n *ShutdownStmt) Accept(v Visitor) (Node, bool) {
...
+	case BRIEOptionRateLimit:
+		return "RATE_LIMIT"
+	case BRIEOptionConcurrency:
+		return "CONCURRENCY"
+	case BRIEOptionChecksum:
+		return "CHECKSUM"
+	case BRIEOptionSendCreds:
+		return "SEND_CREDENTIALS_TO_TIKV"
+	case BRIEOptionBackupTimeAgo:
+		return "SNAPSHOT"
```
Comment (3pointer) [inline ast/misc.go]: should we add BRIEOptionBackupTS after [#172](https://github.com/pingcap/br/pull/172) merged

**Comment by GregoryIan** [comment]
@kennytm please resolve conflicts, it seems to be merged soon

File: ast/misc.go (hunk 3 above)
Comment (kennytm) [reply ast/misc.go]: Yes. But we also need to add a hidden config to pass the timezone.

**Comment by kennytm** [comment]
PTAL again @tangenta @tiancaiamao thanks.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**